app/gui/src/presenter/graph/widgets.rs

```diff
@@ -0,0 +1,352 @@
+    widget_queries:  HashMap<ast::Id, QueryData>,
+}
+
+/// Data of the query that
+#[derive(Debug)]
+struct QueryData {
+    node_id:         ViewNodeId,
+    call_expression: ast::Id,
+    method_name:     ImString,
+    arguments:       Vec<ImString>,
+    last_update:     Option<WidgetUpdates>,
+}
+
+/// Definition of a widget request. Defines the operation expression that the widgets will be
+/// attached to.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct Request {
+    /// The node id of a node that contains the operation expression.
+    pub node_id:           ViewNodeId,
+    /// Expression of the whole method call.
+    pub call_expression:   ast::Id,
+    /// Expression of the call target. Used as a visualization target.
+    pub target_expression: ast::Id,
+    /// The suggestion id of the method that this call refers to.
+    pub call_suggestion:   SuggestionId,
+}
+
+impl Model {
+    fn handle_notification(
+        &mut self,
+        notification: Notification,
+    ) -> Option<(ViewNodeId, WidgetUpdates)> {
+        match notification {
+            Notification::ValueUpdate { target, data, .. } => {
+                let query_data = self.widget_queries.get_mut(&target)?;
+                let args: Vec<(String, serde_json::Value)> = serde_json::from_slice(&data).ok()?;
+                let updates = args
+                    .into_iter()
+                    .map(|(argument_name, meta_json)| {
+                        // Treat failed deserialization as non-widget data. The returned annotation
+                        // can be an arbitrary structure, so there will be cases when we fail to
+                        // deserialize it, as it may not contain the widget data in the first place.
+                        let deserialized = serde_json::from_value(meta_json).ok();
+                        let meta = deserialized.map(WidgetVisualizationData::into_metadata);
+                        WidgetUpdate { argument_name, meta }
+                    })
+                    .collect();
+
+
+                let updates = WidgetUpdates { target_id: target, updates: Rc::new(updates) };
+                trace!("Widget updates: {updates:?}");
+                query_data.last_update = Some(updates.clone());
+                Some((query_data.node_id, updates))
+            }
+            Notification::FailedToAttach { error, .. } => {
+                error!("Failed to attach widget visualization: {error}");
+                None
+            }
+            Notification::FailedToDetach { error, .. } => {
+                error!("Failed to detach widget visualization: {error}");
+                None
+            }
+            Notification::FailedToModify { error, .. } => {
+                error!("Failed to modify widget visualization: {error}");
+                None
+            }
+        }
+    }
+
+    fn request_widget(&mut self, req: &Request) -> Option<(ViewNodeId, WidgetUpdates)> {
```

Comment:
a really long function. Please think if you can split it into several well-named functions and add docs what it does. (if you split it to really well named functions, docs may not be needed).